src/commons/eventsourcing/store.rs

```diff
@@ -417,99 +281,86 @@ where
+                    // The command contains some effect.
+                    let processed_command = processed_command_builder.finish_with_events(events);
 
-                    // Using a lock on the hashmap here to ensure that all updates happen sequentially.
-                    // It would be better to get a lock only for this specific aggregate. So it may be
-                    // worth rethinking the structure.
-                    //
-                    // That said.. saving and applying events is really quick, so this should not hurt
-                    // performance much.
+                    // We will need to apply the command first because:
+                    // a) then we are really, really, sure that it can be applied (no panics)
+                    // b) more importantly, we will need to pass an updated aggregate to pre-save listeners
                     //
-                    // Also note that we don't need the lock to update the inner arc in the cache. We
-                    // just need it to be in scope until we are done updating.
-                    let mut cache = self.cache.write().unwrap();
-
-                    // It should be impossible to get events for the wrong aggregate, and the wrong
-                    // versions, because we are doing the update here inside the outer lock, and aggregates
-                    // generally do not lie about who do they are.
-                    //
-                    // Still.. some defensive coding in case we do have some issue. Double check that the
-                    // events are for this aggregate, and are a contiguous sequence of version starting with
-                    // this version.
-                    let version_before = agg.version();
-                    let nr_events = events.len() as u64;
-
-                    // Event numbers apply to the current version of an aggregate, so the first event
-                    // here applies to the current version (before applying) and the 2nd to +1 and so
-                    // on.
-                    info.last_event = version_before + nr_events - 1;
-
-                    for i in 0..nr_events {
-                        let event = &events[i as usize];
-                        let expected_version = version_before + i;
-                        if event.version() != expected_version || event.handle() != &handle {
-                            error!("Unexpected event: {}", event);
-                            return Err(A::Error::from(AggregateStoreError::WrongEventForAggregate(
-                                handle,
-                                event.handle().clone(),
-                                expected_version,
-                                event.version(),
-                            )));
+                    // Unfortunately, this means that we will need to clone the command.
+                    let agg = Arc::make_mut(&mut latest);
```

Comment:
Same question about `make_mut()` as above.

Reply:
This should work. It was not changed but moved around.

That said, see issue https://github.com/NLnetLabs/krill/issues/1076. We need to rethink how we do transactions, and this will affect how we use the cache.